src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+
+    divisor = lu(2 * I - p_star_X)
+    q .= p * (divisor \ add_scaled_I!(p_star_X, 2.0))
+    return q
+end
+
+Base.show(io::IO, ::Symplectic{n,𝔽}) where {n,𝔽} = print(io, "Symplectic{$(2n), $(𝔽)}()")
+
+function symplectic_inverse_times(M::Symplectic{n}, p, q) where {n}
+    A = similar(p)
+    return symplectic_inverse_times!(M, A, p, q)
+end
+
+function symplectic_inverse_times!(::Symplectic{n}, A, p, q) where {n}
+    # we write p = [p1 p2; p3 p4] (and q, too), then
+    p1 = @view(p[1:n, 1:n])
+    p2 = @view(p[1:n, (n + 1):(2n)])
+    p3 = @view(p[(n + 1):(2n), 1:n])
+    p4 = @view(p[(n + 1):(2n), (n + 1):(2n)])
+    q1 = @view(q[1:n, 1:n])
+    q2 = @view(q[1:n, (n + 1):(2n)])
+    q3 = @view(q[(n + 1):(2n), 1:n])
+    q4 = @view(q[(n + 1):(2n), (n + 1):(2n)])
+    A1 = @view(A[1:n, 1:n])
+    A2 = @view(A[1:n, (n + 1):(2n)])
+    A3 = @view(A[(n + 1):(2n), 1:n])
+    A4 = @view(A[(n + 1):(2n), (n + 1):(2n)])
+    mul!(A1, p4', q1) # A1 = p4'q1
+    mul!(A1, p2', q3, -1, 1) # A1 -= p2'p3
+    mul!(A2, p4', q2) # A2 = p4'q2
+    mul!(A2, p2', q4, -1, 1) #A2 -= p2'q4
+    mul!(A3, p1', q3) #A3 = p1'q3
+    mul!(A3, p3', q1, -1, 1) # A3 -= p3'q1
+    mul!(A4, p1', q4) # A4 = p1'q4
+    mul!(A4, p3', q2, -1, 1) #A4 -= p3'q2
+    return A
+end
+
+ndims(Q::SymplecticMatrix) = 2
+copy(Q::SymplecticMatrix) = SymplecticMatrix(Q.λ)
+Base.eltype(::SymplecticMatrix{T}) where {T} = T
+function Base.convert(::Type{SymplecticMatrix{T}}, Q::SymplecticMatrix) where {T}
+    return SymplecticMatrix(convert(T, Q.λ))
+end
+
+function Base.show(io::IO, Q::SymplecticMatrix)
+    s = "$(Q.λ)"
+    if occursin(r"\w+\s*[\+\-]\s*\w+", s)
+        s = "($s)"
+    end
+    return print(io, typeof(Q), "(): $(s)*[0 I; -I 0]")
+end
+
+(Base.:-)(Q::SymplecticMatrix) = SymplecticMatrix(-Q.λ)
+
+function (Base.:^)(Q::SymplecticMatrix, n::Integer)
+    return ifelse(
+        n % 2 == 0,
+        UniformScaling((-1)^(div(n, 2)) * (Q.λ)^n),
+        SymplecticMatrix((-1)^(div(n - 1, 2)) * (Q.λ)^n),
+    )
+end
+
+(Base.:*)(x::Number, Q::SymplecticMatrix) = SymplecticMatrix(x * Q.λ)
+(Base.:*)(Q::SymplecticMatrix, x::Number) = SymplecticMatrix(x * Q.λ)
+function (Base.:*)(Q1::SymplecticMatrix, Q2::SymplecticMatrix)
+    return LinearAlgebra.UniformScaling(-Q1.λ * Q2.λ)
+end
+
+Base.transpose(Q::SymplecticMatrix) = -Q
+Base.adjoint(Q::SymplecticMatrix) = -SymplecticMatrix(conj(Q.λ))
+Base.inv(Q::SymplecticMatrix) = SymplecticMatrix(-(1 / Q.λ))
+
+(Base.:+)(Q1::SymplecticMatrix, Q2::SymplecticMatrix) = SymplecticMatrix(Q1.λ + Q2.λ)
+(Base.:-)(Q1::SymplecticMatrix, Q2::SymplecticMatrix) = SymplecticMatrix(Q1.λ - Q2.λ)
+
+(Base.:+)(Q::SymplecticMatrix, p::AbstractMatrix) = p + Q
+function (Base.:+)(p::AbstractMatrix, Q::SymplecticMatrix)
+    # When we are adding, the Matrices must match in size:
+    n, _ = get_half_dims(p, true, false, true)
+
+    # Allocate new memory:
+    TS = Base._return_type(+, Tuple{eltype(p),eltype(Q)})
```

Comment:
```suggestion
    TS = typeof(one(eltype(p)) + one(eltype(Q)))
```
This is a more idiomatic element type calculation in Julia. Constant propagation should handle this.